topdown/topdown_test.go

```diff
@@ -189,34 +190,235 @@ func TestTopDownQueryCancellation(t *testing.T) {
+	for i := 0; i < 1000; i++ {
+		arr[i] = i
+	}
 	data := map[string]interface{}{
-		"arr": make([]interface{}, 1000),
+		"arr": arr,
 	}
 
 	store := inmem.NewFromObject(data)
 	txn := storage.NewTransactionOrDie(ctx, store)
 	cancel := NewCancel()
+	buf := NewBufferTracer()
 
 	query := NewQuery(ast.MustParseBody("data.test.p")).
 		WithCompiler(compiler).
 		WithStore(store).
 		WithTransaction(txn).
-		WithCancel(cancel)
+		WithCancel(cancel).
+		WithTracer(buf)
 
+	done := make(chan struct{})
 	go func() {
 		time.Sleep(time.Millisecond * 50)
 		cancel.Cancel()
+		close(done)
 	}()
+	<-done
 
 	qrs, err := query.Run(ctx)
 	if err == nil || err.(*Error).Code != CancelErr {
-		t.Fatalf("Expected cancel error but got: %v (err: %v)", qrs, err)
+		t.Errorf("Expected cancel error but got: %v (err: %v)", qrs, err)
+		PrettyTrace(os.Stdout, []*Event(*buf))
 	}
+}
 
+func TestTopDownEarlyExit(t *testing.T) {
+	// NOTE(sr): There are two ways to early-exit: don't evaluate subsequent
+	// rule bodies, like
+	//
+	//  p {
+	//    true
+	//  }
+	//  p {
+	//    # not evaluated
+	//  }
+	//
+	// and not evaluating subsequent "rounds" of iteration:
+	//
+	//  p {
+	//    x[_] = "y"
+	//  }
+	//
+	// The first kind is relatively simple to test, by having the second rule
+	// body trace something:
+	tests := []struct {
+		note   string
+		module string
+	}{
+		{
+			note: "complete doc",
+			module: `
+				package test
+				p { trace("first body") }
+				p { trace("second body") }`,
+		},
+		{
+			note: "function",
+			module: `
+				package test
+				p = f(1)
+				f(_) { trace("first body") }
+				f(_) { trace("second body") }`,
+		},
+	}
+	t.Run("second rule body not evaluated", func(t *testing.T) {
+		for _, tc := range tests {
+			t.Run(tc.note, func(t *testing.T) {
+				ctx := context.Background()
+				compiler := compileModules([]string{tc.module})
+
+				store := inmem.NewFromObject(map[string]interface{}{})
+				txn := storage.NewTransactionOrDie(ctx, store)
+				buf := NewBufferTracer()
+
+				query := NewQuery(ast.MustParseBody("data.test.p")).
+					WithCompiler(compiler).
+					WithStore(store).
+					WithTransaction(txn).
+					WithTracer(buf)
+
+				_, err := query.Run(ctx)
+				if err != nil {
+					t.Fatalf("Unexpected error: %v", err)
+				}
+				notes := map[string]int{}
+				exits := map[string]int{}
+
+				for _, ev := range []*Event(*buf) {
+					switch ev.Op {
+					case NoteOp:
+						notes[ev.Message]++
+					case ExitOp:
+						exits[ev.Message]++
+					}
+				}
+				if exp, act := 1, len(notes); exp != act {
+					t.Errorf("expected %d note traces, got %d", exp, act)
+					PrettyTrace(os.Stderr, *buf)
+				}
+				if exp, act := 1, exits["early"]; exp != act {
+					t.Errorf("expected %d early exit events, got %d", exp, act)
+					PrettyTrace(os.Stderr, *buf)
+				}
+			})
+		}
+	})
+
+	// For the iteration cancellation, we'll set up a rule body that would take
+	// longer than a certain time budget if it was evaluated in full.
```

Comment:
nit: why can't we use `trace()` again here?

```
p { q[_]; trace("x") }
q[1]
q[2]
```

This should only emit a single trace() with early-exit enabled.

Just bringing this up because anything dependent on time/sleep always bites us eventually.

Reply:
Hah that's a good point. When looking at the traces before, I had only been not finding traces related to the data itertation, but this should do!